Apps/Sandcastle/gallery/Drawing on Terrain.html

```diff
@@ -0,0 +1,185 @@
+            // This is the point that'll follow the mouse.
+            floatingPoint = createPoint(earthPosition);
+            // For the first point, we want to push an extra position into our list of vertices
+            // this will be a dynamic point that'll be replaced when the mouse moves.
+            activeShapePoints.push(earthPosition);
+            // We create a callback property for our positions since we want the shape to update
+            // in real time as we move the mouse. We make sure to pass `false` to the
+            // `CallbackProperty` constructor to mark it as not constant.
+            var dynamicPositions = new Cesium.CallbackProperty(function () {
+                // This is a work-around since polylines when `clampToGround` is true will
+                // crash if you try to draw a zero length line
+                if (activeShapePoints.length === 2) {
+                    var distance = Cesium.Cartesian3.distance(activeShapePoints[0], activeShapePoints[1]);
+                    if (distance === 0) {
+                        return [];
+                    }
+                }
+                return activeShapePoints;
+            }, false);
+            if (drawingMode === 'line') {
+                activeShape = viewer.entities.add({
+                    polyline : {
+                        positions : dynamicPositions,
+                        clampToGround : true,
+                        width : 3
+                    }
+                });
+            }
+            if (drawingMode === 'polygon') {
+                activeShape = viewer.entities.add({
+                    polygon: {
+                        hierarchy: dynamicPositions,
+                        outline: true,
+                        material: new Cesium.ColorMaterialProperty(Cesium.Color.WHITE.withAlpha(0.7))
+                    }
+                });
+            }
+        }
+        // Add this point to the list of vertices defining the shape
+        activeShapePoints.push(earthPosition);
+        // Render a point to mark this vertex
+        createPoint(earthPosition);
+    }
+}, Cesium.ScreenSpaceEventType.LEFT_CLICK);
+// On mouse move, if we have an active shape and a valid point, update the list of points
+// and since the shape is using `CallbackProperty` property it will automatically update.
+handler.setInputAction(function(event) {
+    if (Cesium.defined(floatingPoint)) {
+        var newPosition = viewer.scene.pickPosition(event.endPosition);
+        if (Cesium.defined(newPosition)) {
+            floatingPoint.position.setValue(newPosition);
+            activeShapePoints.pop();
+            activeShapePoints.push(newPosition);
+        }
+    }
+}, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
+// On right click, terminate active shape and redraw it as a new entity without the `CallbackProperty`.
+function terminateShape() {
+    activeShapePoints.pop();
+    if (drawingMode === 'line') {
```

Comment:
This looks duplicated with  L88-105.   Maybe make a function to handle this like you did for `createPoint`